File: internal/dis/pow.go
// internal/dis/pow.go — hashcash-style proof-of-work gate for AdModePOWCAPTCHA.
//
// golden.md Part 3 calls for a "PoW + CAPTCHA gate" in attack mode. A real
// third-party CAPTCHA (reCAPTCHA etc.) would violate the spec's own axioms —
// "no third-party beacons," "no SaaS umbilical cords" — so the gate is pure
// PoW: the browser must find a nonce whose hash (seed || nonce) has at least
// Difficulty leading zero bits before the real ad is served. Cheap for one
// legitimate visitor, expensive to repeat at bot scale.
//
// Stateless by design (no server-side challenge store, matching "no
// persistent identifiers" and the ring buffer's local-first footprint): the
// challenge token embeds its own expiry and an HMAC over (seed, expiry)
// under a secret generated once at collector startup. Verify just recomputes
// the HMAC and checks expiry + proof — no map, no cleanup goroutine, no
// growth under load.
package dis

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"strconv"
	"time"
)

// DefaultPOWDifficulty is the number of required leading zero bits.
// ~2^16 average hash attempts — a few hundred ms to a couple seconds via
// crypto.subtle in a browser, negligible for one human, real cost at the
// request volumes an attack posture implies.
const DefaultPOWDifficulty = 16

// POWChallengeTTL is how long an issued challenge remains valid.
const POWChallengeTTL = 60 * time.Second

var errBadToken = errors.New("dis: malformed pow token")

// POWIssuer issues and verifies stateless PoW challenges.
type POWIssuer struct {
	secret     [32]byte
	Difficulty uint8
}

// NewPOWIssuer generates a fresh random secret (instance-local; a collector
// restart invalidates outstanding challenges, which is fine — they're only
// ever valid for POWChallengeTTL anyway).
func NewPOWIssuer(difficulty uint8) (*POWIssuer, error) {
	iss := &POWIssuer{Difficulty: difficulty}
	if _, err := rand.Read(iss.secret[:]); err != nil {
		return nil, err
	}
	return iss, nil
}

// Challenge is what's handed to the client.
type Challenge struct {
	Seed       string `json:"seed"`       // hex-encoded random seed the client hashes against
	Token      string `json:"token"`      // opaque, verify passes this back unchanged
	Difficulty uint8  `json:"difficulty"` // required leading zero bits
	ExpiresAt  int64  `json:"expires_at"` // unix seconds
}

// Issue returns a new Challenge. No state is retained server-side.
func (p *POWIssuer) Issue(now time.Time) (Challenge, error) {
	var seed [16]byte
	if _, err := rand.Read(seed[:]); err != nil {
		return Challenge{}, err
	}
	expires := now.Add(POWChallengeTTL).Unix()
	token := p.sign(seed[:], expires)
	return Challenge{
		Seed:       encodeHex(seed[:]),
		Token:      token,
		Difficulty: p.Difficulty,
		ExpiresAt:  expires,
	}, nil
}

// sign builds the opaque token: base64( seed || expiry(8 bytes BE) || hmac ).
func (p *POWIssuer) sign(seed []byte, expires int64) string {
	var expBuf [8]byte
	binary.BigEndian.PutUint64(expBuf[:], uint64(expires))

	mac := hmac.New(sha256.New, p.secret[:])
	mac.Write(seed)
	mac.Write(expBuf[:])
	sum := mac.Sum(nil)

	buf := make([]byte, 0, len(seed)+len(expBuf)+len(sum))
	buf = append(buf, seed...)
	buf = append(buf, expBuf[:]...)
	buf = append(buf, sum...)
	return base64.RawURLEncoding.EncodeToString(buf)
}

// Verify checks the token's authenticity/expiry, then checks that nonce is a
// valid proof-of-work solution for the token's embedded seed.
func (p *POWIssuer) Verify(token, nonce string, now time.Time) (bool, error) {
	raw, err := base64.RawURLEncoding.DecodeString(token)
	if err != nil {
		return false, errBadToken
	}
	// seed(16) + expiry(8) + hmac-sha256(32)
	if len(raw) != 16+8+sha256.Size {
		return false, errBadToken
	}
	seed := raw[:16]
	expBuf := raw[16:24]
	gotMAC := raw[24:]

	expires := int64(binary.BigEndian.Uint64(expBuf))

	mac := hmac.New(sha256.New, p.secret[:])
	mac.Write(seed)
	mac.Write(expBuf)
	wantMAC := mac.Sum(nil)

	if subtle.ConstantTimeCompare(gotMAC, wantMAC) != 1 {
		return false, nil // tampered or issued by a different collector instance
	}
	if now.Unix() > expires {
		return false, nil // expired — client must request a fresh challenge
	}
	return checkProof(encodeHex(seed), nonce, p.Difficulty), nil
}

// checkProof reports whether sha256(seedHex || nonce) has at least
// difficulty leading zero bits.
func checkProof(seedHex, nonce string, difficulty uint8) bool {
	h := sha256.Sum256([]byte(seedHex + nonce))
	return leadingZeroBits(h[:]) >= difficulty
}

func leadingZeroBits(b []byte) uint8 {
	var n uint8
	for _, byte_ := range b {
		if byte_ == 0 {
			n += 8
			continue
		}
		for mask := byte(0x80); mask > 0; mask >>= 1 {
			if byte_&mask != 0 {
				return n
			}
			n++
		}
	}
	return n
}

func encodeHex(b []byte) string {
	const hexDigits = "0123456789abcdef"
	out := make([]byte, len(b)*2)
	for i, c := range b {
		out[i*2] = hexDigits[c>>4]
		out[i*2+1] = hexDigits[c&0x0f]
	}
	return string(out)
}

// findProof is a reference solver used only by tests — the real solving
// happens client-side in JS. Kept here so pow_test.go can generate a valid
// nonce without duplicating the hashing logic.
func findProof(seedHex string, difficulty uint8) string {
	for i := 0; ; i++ {
		nonce := strconv.Itoa(i)
		if checkProof(seedHex, nonce, difficulty) {
			return nonce
		}
	}
}

File: plugins/edis-dis/assets/pow.js
/**
 * EDIS DIS — client-side proof-of-work solver for the AdModePOWCAPTCHA gate.
 *
 * One legitimate visitor pays this cost once per challenge (a few hundred ms
 * to ~1-2s at the default difficulty). A bot re-running it at attack volume
 * pays it every single time — that asymmetry is the whole defensive value.
 * No third-party CAPTCHA service is involved (see golden.md's own axioms:
 * "no third-party beacons," "no SaaS umbilical cords") — this is pure
 * self-hosted hashcash, verified back against the DIS collector.
 */
(function () {
	'use strict';

	function toHex(buffer) {
		var bytes = new Uint8Array(buffer);
		var hex = '';
		for (var i = 0; i < bytes.length; i++) {
			hex += bytes[i].toString(16).padStart(2, '0');
		}
		return hex;
	}

	// Leading zero *bits* of a hex-encoded digest, matching the Go server's
	// bit-level check exactly (not just whole hex-nibble zeros).
	function leadingZeroBits(hex) {
		var bits = 0;
		for (var i = 0; i < hex.length; i++) {
			var nibble = parseInt(hex[i], 16);
			if (nibble === 0) {
				bits += 4;
				continue;
			}
			// Count leading zero bits within this nibble.
			if (nibble < 8) bits += 1;
			if (nibble < 4) bits += 1;
			if (nibble < 2) bits += 1;
			return bits;
		}
		return bits;
	}

	async function sha256Hex(input) {
		var data = new TextEncoder().encode(input);
		var digest = await crypto.subtle.digest('SHA-256', data);
		return toHex(digest);
	}

	async function solve(seed, difficulty, deadlineMs) {
		var start = performance.now();
		for (var i = 0; ; i++) {
			var nonce = String(i);
			var hex = await sha256Hex(seed + nonce);
			if (leadingZeroBits(hex) >= difficulty) {
				return nonce;
			}
			if (performance.now() - start > deadlineMs) {
				throw new Error('pow solve timeout');
			}
		}
	}

	async function runGate(el) {
		var restUrl = window.edisDisPow && window.edisDisPow.restUrl;
		var nonceHeader = window.edisDisPow && window.edisDisPow.nonce;
		if (!restUrl) return;

		var slot = el.getAttribute('data-slot') || 'default';
		var text = el.getAttribute('data-text') || '';
		var href = el.getAttribute('data-href') || '';

		try {
			var challengeResp = await fetch(restUrl + 'dis-pow-challenge', {
				headers: { 'X-WP-Nonce': nonceHeader },
			});
			if (!challengeResp.ok) throw new Error('challenge fetch failed');
			var challenge = await challengeResp.json();

			var nonce = await solve(challenge.seed, challenge.difficulty, 8000);

			var verifyResp = await fetch(restUrl + 'dis-pow-verify', {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					'X-WP-Nonce': nonceHeader,
				},
				body: JSON.stringify({
					token: challenge.token,
					nonce: nonce,
					slot: slot,
					text: text,
					href: href,
				}),
			});
			if (!verifyResp.ok) throw new Error('verify failed');
			var result = await verifyResp.json();
			if (result.ok && result.html) {
				el.outerHTML = result.html;
			} else {
				el.remove(); // fail closed for this slot only — never blocks the page
			}
		} catch (e) {
			// Fail closed on the ad slot, never on the page. Attack-mode traffic
			// that can't/won't run the PoW simply doesn't see an ad — correct
			// per golden.md 3.7 ("never block content").
			el.remove();
		}
	}

	document.addEventListener('DOMContentLoaded', function () {
		var gates = document.querySelectorAll('.edis-ad--challenge');
		gates.forEach(runGate);
	});
})();
